### PR TITLE
Run for dan

### PR DESCRIPTION
**Pull request by dmacmillan**
I think that I have refined this branch well enough that it is ready for commit. In addition to my git commit messages that specify the changes I've made, I have also made sure that all the tests pass (which involved installation of some additional tools such as `merge-mates` and IVA (Sanger).

This PR addresses [this issue](https://github.com/cfe-lab/MiCall/issues/473), please let me know if I have missed anything! Feel free to annotate the changes with comments!

### CODE REVIEW

**Comment by dmacmillan** [comment]
It looks like though the tests passed, some succeeding commands failed, I will look into it

File: micall/monitor/kive_watcher.py [hunk 1]
```diff
@@ -263,7 +263,7 @@ def calculate_retry_wait(min_wait, max_wait, attempt_count):
 class KiveWatcher:
     def __init__(self,
                  config=None,
-                 result_handler=lambda result_folder: None,
+                 qai_upload_queue=Queue(),
```
Comment (donkirkby) [inline micall/monitor/kive_watcher.py]: A [mutable value] for a default parameter is almost always a bad idea.

[mutable value]: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

File: micall/monitor/kive_watcher.py [hunk 2]
```diff
@@ -485,7 +485,8 @@ def poll_runs(self):
                     if results_path is None:
                         continue
                     if (results_path / "coverage_scores.csv").exists():
-                        self.result_handler(results_path)
+                        # self.result_handler(results_path)
```
Comment (donkirkby) [inline micall/monitor/kive_watcher.py]: I don't think we need this commented code anymore.

File: micall/monitor/qai_helper.py [hunk 3]
```diff
@@ -15,9 +15,12 @@ def login(self, qai_path, qai_user, password):
...
 
-        response = self.post(qai_path + "/account/login",
-                             data={'user_login': qai_user,
-                                   'user_password': password})
+        try:
+            response = self.post(qai_path + "/account/login",
+                                data={'user_login': qai_user,
+                                    'user_password': password})
+        except Exception as e:
+            raise RuntimeError('Could not log in to QAI') from None
```
Comment (donkirkby) [inline micall/monitor/qai_helper.py]: Why hide the original exception? Consider these options:

1. Don't catch the exception here. Just let the original exception bubble up to the upload loop and get logged there.
2. If the original exception doesn't have enough information to make it clear that it happened while trying to log in to QAI, just remove the `from None`. That will log both the original exception and the fact that it was trying to log in to QAI. You can be even more explicit about the link between the two exceptions by using `from e`.
3. If there is something in the original exception that was harmful or confusing to log, add a comment to explain why you added the `from None`.

In option 2 or 3, you should probably be more specific about the exception that you're catching. At least `RequestException`, and maybe `ConnectionError`.

File: micall/monitor/update_qai.py [hunk 4]
```diff
@@ -370,29 +374,60 @@ def process_folder(result_folder,
...
+                                        f3,
+                                        run,
+                                        sample_sheet,
+                                        conseqs,
+                                        session,
+                                        pipeline_version)
+                    logger.info('Upload success!')
+                    break
+            except Exception as e:
+                logger.error('Upload to QAI FAILED')
```
Comment (donkirkby) [inline micall/monitor/update_qai.py]: Add the `exc_info=True` parameter, so you get the details of why it failed.

File: micall/monitor/qai_helper.py (hunk 3 above)
Comment (dmacmillan) [reply micall/monitor/qai_helper.py]: The traceback of the exception that is caught at this stage is quite verbose and most of it isn't really helpful, hence why I opted for this method (especially if it's part of a loop structure). I think the best option would be number 3 here, though I agree I could make the exception more specific. 

To resolve this I will:

1. Add a comment explaining `from None` was added to reduce the log message size which involves several nested exceptions being handled
2. Replace the `Exception` with `ConnectionError`

File: micall/monitor/kive_watcher.py (hunk 1 above)
Comment (dmacmillan) [reply micall/monitor/kive_watcher.py]: Good catch

**Comment by dmacmillan** [comment]
Implemented the requested changes from the last PR

File: micall/monitor/qai_helper.py (hunk 3 above)
Comment (donkirkby) [reply micall/monitor/qai_helper.py]: I was suggesting that you *catch* a more specific exception class. Throwing the `RuntimeError` is fine. The rest of your plan sounds good to me.

Comment (dmacmillan) [reply micall/monitor/qai_helper.py]: I see, I've replaced it with `RequestException`

**Pull request merged**